core/src/main/java/hudson/PluginManager.java

```diff
@@ -137,7 +137,25 @@
     /**
      * All active plugins, topologically sorted so that when X depends on Y, Y appears in the list before X does.
      */
-    protected final List<PluginWrapper> activePlugins = new CopyOnWriteArrayList<PluginWrapper>();
+    protected final List<PluginWrapper> activePlugins = new CopyOnWriteArrayList<PluginWrapper>() {
+        @Override
+        public boolean add(PluginWrapper pluginWrapper) {
+            try {
+                return super.add(pluginWrapper);
+            } finally {
+                onActivate(pluginWrapper);
+            }
+        }
+
+        @Override
+        public boolean remove(Object pluginWrapper) {
+            try {
+                return super.remove(pluginWrapper);
+            } finally {
+                onDeactivate((PluginWrapper) pluginWrapper);
+            }
+        }
+    };
```

Comment:
Some people might object to triggering the events this way. The `onActivate` could also be done at the end of the `dynamicLoad` method. Can easily move there if people object.